presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java

```diff
@@ -0,0 +1,211 @@
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_BAD_CREDENTIALS_ERROR, e);
+        }
+        long expiresAfterWriteMillis = config.getSheetsDataExpireAfterWrite().toMillis();
+        long maxCacheSize = config.getSheetsDataMaxCacheSize();
+
+        this.tableSheetMappingCache = newCacheBuilder(expiresAfterWriteMillis, maxCacheSize)
+                .build(new CacheLoader<String, Optional<String>>() {
+                    @Override
+                    public Optional<String> load(String tableName)
+                    {
+                        return getSheetExpressionForTable(tableName);
+                    }
+
+                    @Override
+                    public Map<String, Optional<String>> loadAll(Iterable<? extends String> tableList)
+                    {
+                        return getSheetExpressionForAllTables();
+                    }
+                });
+
+        this.sheetDataCache = newCacheBuilder(expiresAfterWriteMillis, maxCacheSize).build(from(this::getDataForSheetExpression));
+    }
+
+    private Optional<String> getSheetExpressionForTable(String tableName)
+    {
+        try {
+            List<List<Object>> data = readAllValuesFromSheetExpression(metadataSheetId);
+            // first line is assumed to be sheet header
+            for (int i = 1; i < data.size(); i++) {
+                if (data.get(i).size() >= 2) {
+                    String tableId = String.valueOf(data.get(i).get(0));
+                    String sheetId = String.valueOf(data.get(i).get(1));
+                    if (tableName.equalsIgnoreCase(tableId)) {
+                        return Optional.of(sheetId);
+                    }
+                }
+            }
+        }
+        catch (IOException e) {
+            throw new PrestoException(SHEETS_METASTORE_ERROR, e);
+        }
+        return Optional.empty();
+    }
+
+    private Map<String, Optional<String>> getSheetExpressionForAllTables()
+    {
+        ImmutableMap.Builder<String, Optional<String>> tableSheetMap = ImmutableMap.builder();
+        try {
+            List<List<Object>> data = readAllValuesFromSheetExpression(metadataSheetId);
+            // first line is assumed to be sheet header
+            for (int i = 1; i < data.size(); i++) {
+                if (data.get(i).size() >= 2) {
+                    String tableId = String.valueOf(data.get(i).get(0));
+                    String sheetId = String.valueOf(data.get(i).get(1));
+                    tableSheetMap.put(tableId, Optional.of(sheetId));
+                }
+            }
+        }
+        catch (IOException e) {
+            throw new PrestoException(SHEETS_METASTORE_ERROR, e);
+        }
+        return tableSheetMap.build();
+    }
+
+    private Optional<List<List<Object>>> getDataForSheetExpression(String sheetExpression)
```

Comment:
this method does not return `Optional.empty()` so it `Optional` is not needed here